src/main/java/org/truffleruby/core/rope/RopeNodes.java

```diff
@@ -242,6 +244,140 @@ protected static boolean is7Bit(CodeRange codeRange) {
+                                                                   @Cached("create()") BranchProfile brokenCodeRangeProfile) {
+            // Taken from StringSupport.strLengthWithCodeRangeAsciiCompatible.
+
+            CodeRange codeRange = CR_7BIT;
+            int characters = 0;
+            int p = 0;
+            final int end = bytes.length;
+
+            while (p < end) {
+                if (Encoding.isAscii(bytes[p])) {
+                    final int multiByteCharacterPosition = StringSupport.searchNonAscii(bytes, p, end);
+                    if (multiByteCharacterPosition == -1) {
+                        return new StringAttributes(characters + (end - p), codeRange);
+                    }
+
+                    characters += multiByteCharacterPosition - p;
+                    p = multiByteCharacterPosition;
+                }
+
+                final int lengthOfCurrentCharacter = StringSupport.preciseLength(encoding, bytes, p, end);
+
+                if (lengthOfCurrentCharacter > 0) {
+                    multiByteCharacterProfile.enter();
+
+                    if (codeRange != CR_BROKEN) {
+                        codeRange = CR_VALID;
+                    }
+
+                    p += lengthOfCurrentCharacter;
+                } else {
+                    brokenCodeRangeProfile.enter();
+                    codeRange = CR_BROKEN;
+                    p++;
+                }
+
+                characters++;
+            }
+
+            return new StringAttributes(characters, codeRange);
+        }
+
+        @Specialization(guards = { "!isEmpty(bytes)", "!isBinaryString(encoding)", "!isAsciiCompatible(encoding)" })
+        StringAttributes calculateAttributesGeneric(Encoding encoding, byte[] bytes,
+                                                    @Cached("createBinaryProfile()") ConditionProfile asciiCompatibleProfile,
+                                                    @Cached("create()") BranchProfile multiByteCharacterProfile,
+                                                    @Cached("create()") BranchProfile brokenCodeRangeProfile) {
+            // Taken from StringSupport.strLengthWithCodeRangeNonAsciiCompatible.
+
+            CodeRange codeRange = asciiCompatibleProfile.profile(encoding.isAsciiCompatible()) ? CR_7BIT : CR_VALID;
+            int characters;
+            int p = 0;
+            final int end = bytes.length;
+
+            for (characters = 0; p < end; characters++) {
+                final int lengthOfCurrentCharacter = StringSupport.preciseLength(encoding, bytes, p, end);
+
+                if (lengthOfCurrentCharacter > 0) {
+                    multiByteCharacterProfile.enter();
+
+                    if (codeRange != CR_BROKEN) {
+                        codeRange = CR_VALID;
+                    }
+
+                    p += lengthOfCurrentCharacter;
+                } else {
+                    brokenCodeRangeProfile.enter();
```

Comment:
A single ConditionProfile would be clearer here since both branches are profiled.